### PR TITLE
[Market] Add market punishing mechanism

### LINKED ISSUE #137: [Market] Storage slashing
we ONLY slash provider. To apply market slashing, we should record provider's `failed time`(which can be easily recorded) and slashing according to this data.

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #137

### CODE REVIEW

**Comment by zikunfan** [comment]
Decide to rename all `params` and `variables` into `punishing/punishment/punish` to distinguished with `slashing` in ***GPoS*** module

File: cstrml/market/src/lib.rs [hunk 1]
```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
...
+        });
+
+        // 2. Maybe do slash
+        Self::maybe_slash_provider(&order_id, &so);
+    }
+
+    pub fn maybe_slash_provider(order_id: &T::Hash, so: &StorageOrder<T::AccountId, BalanceOf<T>>) {
+        let mut status_count = Self::order_slash_records(order_id).unwrap();
+        let slash_duration = T::SlashingDuration::get();
+        if status_count.success + status_count.failed >= slash_duration && status_count.can_slash {
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:491]: Is there a simple way to judge the punishing condition?
Like recording the `current_era`?

File: cstrml/market/src/lib.rs [hunk 2]
```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
...
+            // Calculate slash value
+            let mut slash_value = Zero::zero();
+            if status_count.success >= ((slash_duration as f64) * 0.95) as EraIndex {
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:494]: Using `prebill`

File: cstrml/market/src/lib.rs [hunk 3]
```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
...
+            }
+
+        });
+
+        // 2. Maybe do slash
+        Self::maybe_slash_provider(&order_id, &so);
+    }
+
+    pub fn maybe_slash_provider(order_id: &T::Hash, so: &StorageOrder<T::AccountId, BalanceOf<T>>) {
+        let mut status_count = Self::order_slash_records(order_id).unwrap();
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:489]: `status_count` looks no meaning to me

File: cstrml/market/src/lib.rs [hunk 4]
```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
...
+        let slash_duration = T::SlashingDuration::get();
+        if status_count.success + status_count.failed >= slash_duration && status_count.can_slash {
+            // Calculate slash value
+            let mut slash_value = Zero::zero();
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:493]: Simply use `u64`?

File: cstrml/market/src/lib.rs [hunk 5]
```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
...
+            else {
+                slash_value = so.amount;
+            }
+
+            // Do slash
+            if !slash_value.is_zero() {
+                slash_value = slash_value.min(so.amount - status_count.slashed_value);
+                status_count.slashed_value += slash_value;
+                if status_count.slashed_value >= so.amount {
+                    status_count.can_slash = false;
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:512]: What if the `can_punish` be false? Will this influent delayed payment? like If the `remain_value`(pledge - punished_value) < `sorder_amount`, payment should failed

File: cstrml/payment/src/mock.rs [hunk 6]
```diff
@@ -140,12 +140,19 @@ impl tee::Trait for Test {
...
+
 impl market::Trait for Test {
     type Currency = Balances;
+    type CurrencyToBalance = CurrencyToVoteHandler;
     type Event = ();
     type Randomness = ();
     type Payment = Payment;
     type OrderInspector = TestOrderInspector;
+    type SlashingDuration = SlashingDuration;
+
```
Comment (zikunfan) [inline cstrml/payment/src/mock.rs:155]: extra line

File: cstrml/market/src/lib.rs (hunk 5 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:512]: it won't have impact on delayed payment.

File: cstrml/market/src/lib.rs (hunk 4 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:493]: using perbill and removed type tranformation

File: cstrml/market/src/lib.rs (hunk 3 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:489]: rename it to punishment

File: cstrml/market/src/lib.rs (hunk 2 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:494]: it's just EraIndex and perbill cannot be applied

File: cstrml/market/src/lib.rs [hunk 7]
```diff
@@ -53,6 +56,15 @@ pub enum OrderStatus {
     Pending
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct OrderSlashRecord<Balance> {
+    pub success: EraIndex,
+    pub failed: EraIndex,
+    pub slashed_value: Balance,
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:64]: rename it to punished_value

File: cstrml/market/src/lib.rs [hunk 8]
```diff
@@ -53,6 +56,15 @@ pub enum OrderStatus {
     Pending
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct OrderSlashRecord<Balance> {
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:61]: based on our discussion, rename it to ProviderPunishment

File: cstrml/market/src/lib.rs [hunk 9]
```diff
@@ -429,16 +457,66 @@ impl<T: Trait> Module<T> {
...
+            // 3. reset
+            punishment.success = 0;
+            punishment.failed = 0;
+        } else {
+
+        }
+        <ProviderPunishments<T>>::insert(&order_id, punishment);
+    }
+
+    pub fn do_slash_provider(
```
Comment (MyronFanQiu) [inline cstrml/market/src/lib.rs:492]: using slash here since it's a slash from balance perspective

File: cstrml/market/src/lib.rs (hunk 1 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs]: each order has its own check and cannot have a current_era.

File: cstrml/market/src/lib.rs [hunk 10]
```diff
@@ -53,6 +56,14 @@ pub enum OrderStatus {
     Pending
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct ProviderPunishment<Balance> {
+    pub success: EraIndex,
+    pub failed: EraIndex,
+    pub punished_value: Balance
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:64]: Just named `value`?

File: cstrml/market/src/lib.rs (hunk 9 above)
Comment (zikunfan) [reply cstrml/market/src/lib.rs:492]: No, I think we should unify all ***Code Concept***. Just named `punish_provider` is okay.

Comment (zikunfan) [reply cstrml/market/src/lib.rs:492]: No, I think we should unify all ***Code Concept***. 
Moreover, this function did lots of punishment dealing logic.
As my perspective, just named `punish_provider` is fine.

**Pull request merged**